### ISSUE TITLE
Tkinter checkbutton switch on and off together

### ISSUE THREAD

**Issue by e1bf2393-cc11-4392-9217-69984ffa2a8c** [python/cpython#83079]
BPO | [38898](https://bugs.python.org/issue38898)
--- | :---
Nosy | @serhiy-storchaka
Superseder | <li>bpo-29402: Problem with Checkbutton and  duplicate last name components</li>
Files | <li>[tkinter-checkbutton-bug.ogv](https://bugs.python.org/file48741/tkinter-checkbutton-bug.ogv "Uploaded as video/ogg at 2019-11-22.21:16:19 by zaza hohonini"): Video demonstatring the problem</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2019-11-28.07:04:43.816>
created_at = <Date 2019-11-22.21:16:20.107>
labels = ['3.8', '3.7', 'expert-tkinter']
title = 'Tkinter checkbutton switch on and off together'
updated_at = <Date 2019-11-28.07:04:43.815>
user = 'https://bugs.python.org/zazahohonini'
```

bugs.python.org fields:
```python
activity = <Date 2019-11-28.07:04:43.815>
actor = 'serhiy.storchaka'
assignee = 'none'
closed = True
closed_date = <Date 2019-11-28.07:04:43.816>
closer = 'serhiy.storchaka'
components = ['Tkinter']
creation = <Date 2019-11-22.21:16:20.107>
creator = 'zaza hohonini'
dependencies = []
files = ['48741']
hgrepos = []
issue_num = 38898
keywords = []
message_count = 2.0
messages = ['357323', '357375']
nosy_count = 3.0
nosy_names = ['gpolo', 'serhiy.storchaka', 'zaza hohonini']
pr_nums = []
priority = 'normal'
resolution = 'duplicate'
stage = 'resolved'
status = 'closed'
superseder = '29402'
type = None
url = 'https://bugs.python.org/issue38898'
versions = ['Python 3.6', 'Python 3.7', 'Python 3.8']
```

</p></details>

**Comment by questaware** [comment]
It seems that things have not been resolved.

I have this problem which shows that identification of tk.Checkbutton and ttk.Checkbutton is still not working. (Are there any patches for this?)

`import tkinter as tk
from tkinter import Tk
import tkinter.ttk as ttk

def appendBool(flagList, val):
          res = tk.IntVar()
          res.set(val)
          flagList.append(res)
          return res

def NewCheckbutton(frame, text, myvar, state, row, col):
#         global rowCount
          setit=myvar.get()
          b = tk.Checkbutton(frame, text=text, variable=myvar, offvalue=0, onvalue=1, state=state)
#         b = ttk.Checkbutton(frame, text=text, variable=myvar, offvalue=0, onvalue=1, state=state)
          b.grid(row=row, column=col, sticky="EW", ipadx=1, ipady=1)
          b.select()
          b.toggle()
#         b.invoke()
          root.update_idletasks()
          print(b)
          print(f'VG={setit}')
          if setit:
            b.toggle()
#           b.invoke()

          myvar.set(setit)
          root.update_idletasks()
          return b

def do_next():
      global toprow
      toprow = toprow + 1
      disables = []
      disV = appendBool(disables, 1)
      a = NewCheckbutton(root, text="NeXt", myvar=disV, state="normal", row=toprow, col=0)

      disV = appendBool(disables, 0)
      a = NewCheckbutton(root, text="NeXt", myvar=disV, state="normal", row=toprow, col=1)

      root.update_idletasks()
  
paper = "#e0faff"

if 1:
    toprow=6
    root = Tk()
    root.configure(background=paper)

    btn_go = tk.Button(root, text="Continue", width=30, command=do_next)
    btn_go.grid(row=0, pady=20)

    for row in range(0,3):

      disables = []
      disV = appendBool(disables, 1)
      c = NewCheckbutton(root, text="eX", myvar=disV, state="normal", row=row+1, col=0)

      disV = appendBool(disables, 0)
      d = NewCheckbutton(root, text="eX", myvar=disV, state="normal", row=row+1, col=1)

      root.update_idletasks()

    root.mainloop()
`

**Comment by questaware** [comment]
I am running python 3.10 Tk/Tcl 3.6.10.

It seems that the problem is more widespread.

Adding variable=tk.IntVar() does fix the above problem but there is also this problem:

[cb3.txt](https://github.com/python/cpython/files/9627281/cb3.txt)

**Comment by questaware** [comment]
To be more precise the problem is fixed by making what follows variable= be an expression.
Furthermore it must not be possible for Python to optimise it away.
I use 

          `global checkbuttons, cbix

           cbix += 1

           checkbuttons.append( myvar )

           b = tk.Checkbutton(frame, text=text, variable=checkbuttons[cbix], ...`

This does not work:

          `global checkbuttons, cbix

           checkbuttons = [ myvar ]

           b = tk.Checkbutton(frame, text=text, variable=checkbuttons.pop(), ...`
(It seems that the GitHub formatting language is flakey also!)

**Comment by serhiy-storchaka** [comment]
This issue was closed as a duplicate of #73588.
